### PR TITLE
[web] Support config file

### LINKED ISSUE #427: CodeChecker config file support
it should be possible to configure codechecker from config file.

When many programmers use CodeChecker for analyzing their product, it is a common requirement to run the analysis using the same parameters: checker list, gcc_logger_like parameters etc.

The config file path should be possible to be defined from an optional command line parameter or from and environment variable (e.g. $CODECHECKER_CONFIG) . Thus the same config env can be set up for all developers.

- [x] analyzer #2268
- [ ] web #2611

### PR DESCRIPTION
**Pull request by csordasmarton**
> Closes #427

### CODE REVIEW

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/tests/functional/config/__init__.py  2
- web/tests/functional/config/test_store_config.py  3
- web/tests/functional/config/test_server_config.py  3
         

Clones added
============
- web/client/codechecker_client/cmd/store.py  1
- web/tests/functional/config/test_store_config.py  1
- web/server/codechecker_server/cmd/server.py  1
- web/tests/functional/config/test_server_config.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)

File: docs/web/user_guide.md [hunk 1]
```diff
@@ -293,7 +303,14 @@ optional arguments:
                         over the Internet. (Equivalent to specifying '--host
                         ""'.) (default: False)
   --skip-db-cleanup     Skip performing cleanup jobs on the database like
-                        removing unused files.
+                        removing unused files. (default: False)
+  --config CONFIG_FILE  Allow the configuration from an explicit JSON based
+                        configuration file. The value of the 'server' key in
+                        the config file will be emplaced as command line
+                        arguments. The format of configuration file is: {
```
Comment (gyorb) [inline docs/web/user_guide.md]: Please format the json here too.

File: web/client/codechecker_client/cmd/store.py [hunk 2]
```diff
@@ -143,6 +143,23 @@ def add_arguments_to_parser(parser):
...
+    parser.add_argument('--config',
+                        dest='config_file',
+                        required=False,
+                        help="Allow the configuration from an explicit JSON "
+                             "based configuration file. The value of the "
+                             "'store' key in the config file will be "
+                             "emplaced as command line arguments. The format "
+                             "of configuration file is: "
+                             "{"
+                             "  \"enabled\": true,"
```
Comment (gyorb) [inline web/client/codechecker_client/cmd/store.py]: To me this enabled value is a bit confusing, if I want to use a config file I set it in the command line, if not I do not set it.
But why would I set it in the command line and than disable it in the config file?
@bruntib @dkrupp what do you think?

File: web/client/codechecker_client/cmd/store.py [hunk 3]
```diff
@@ -143,6 +143,23 @@ def add_arguments_to_parser(parser):
                              "If multiple prefix is given, the longest match "
                              "will be removed.")
 
+    parser.add_argument('--config',
+                        dest='config_file',
+                        required=False,
+                        help="Allow the configuration from an explicit JSON "
+                             "based configuration file. The value of the "
```
Comment (gyorb) [inline web/client/codechecker_client/cmd/store.py]: I would probably rephrase this to something like this: the values configured in the config file will overwrite the values set in the command line.

File: web/tests/functional/config/__init__.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
+# coding=utf-8
```
Comment (gyorb) [inline web/tests/functional/config/__init__.py]: We have a lot of config files and values, what do you think about renaming the test directory to `cli_config` or something different which indicates that these are the tests for the cli configuration?

File: web/tests/functional/config/__init__.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    # details.
+    print("This test uses a CodeChecker server... connecting...")
+    server_access = codechecker.start_or_get_server()
+    server_access['viewer_product'] = 'config'
+    codechecker.add_test_package_product(server_access, TEST_WORKSPACE)
+
+    # Extend the checker configuration with the server access.
+    codechecker_cfg.update(server_access)
+
+    ret = codechecker.analyze(codechecker_cfg, test_project_path)
```
Comment (gyorb) [inline web/tests/functional/config/__init__.py]: Could we limit this test only to store and not analyze, we can save a lot of time during the tests if we do not have to run the analysis.

File: web/tests/libtest/codechecker.py [hunk 6]
```diff
@@ -582,6 +582,9 @@ def wait_for_server_start(stdoutfile):
                 out = f.read()
                 if "Server waiting for client requests" in out:
                     return
+                if "usage: CodeChecker" in out:
```
Comment (gyorb) [inline web/tests/libtest/codechecker.py]: Could you add a comment about this is an error case when the server failed to start and gave some error message.

File: web/tests/functional/config/__init__.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    # TODO: If environment variable is set keep the workspace
+    # and print out the path.
+    global TEST_WORKSPACE
+
+    check_env = env.import_test_cfg(TEST_WORKSPACE)[
+        'codechecker_cfg']['check_env']
+    codechecker.remove_test_package_product(TEST_WORKSPACE, check_env)
+
+    print("Removing: " + TEST_WORKSPACE)
+    # shutil.rmtree(TEST_WORKSPACE)
```
Comment (gyorb) [inline web/tests/functional/config/__init__.py]: This should not be commented out right? Who does the cleanup?

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/tests/functional/cli_config/__init__.py  1
- web/tests/functional/cli_config/test_server_config.py  3
- web/tests/functional/cli_config/test_store_config.py  3
- codechecker_common/arg.py  5
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/arg.py  -3
         

Clones added
============
- web/tests/functional/cli_config/test_server_config.py  7
- web/client/codechecker_client/cmd/store.py  1
- web/tests/functional/cli_config/test_store_config.py  1
- web/server/codechecker_server/cmd/server.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)

File: web/client/codechecker_client/cmd/store.py (hunk 2 above)
Comment (csordasmarton) [reply web/client/codechecker_client/cmd/store.py]: The [`.codechecker.passwords.json`](https://github.com/Ericsson/codechecker/blob/master/docs/web/authentication.md#preconfigured-credentials-) and the [`server_config.json`](https://github.com/Ericsson/codechecker/blob/master/web/server/config/server_config.json) look the same: both have an option to turn off or on the feature. I wanted to be more consistent with these features. By the way the analyzer --config option works the same so if we would like to change this behaviour we can do it in another pull request. Maybe we can create an issue for this.

File: web/tests/functional/cli_config/__init__.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
+# coding=utf-8
```
Comment (gyorb) [inline web/tests/functional/cli_config/__init__.py]: the coding part is not needed by default in py3 every file should be utf-8

File: web/client/codechecker_client/cmd/store.py (hunk 2 above)
Comment (gyorb) [reply web/client/codechecker_client/cmd/store.py]: I've created a separate issue to discuss it #2654

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- codechecker_common/arg.py  5
- web/tests/functional/cli_config/test_store_config.py  3
- web/tests/functional/cli_config/test_server_config.py  3
- web/tests/functional/cli_config/__init__.py  1
         

Complexity decreasing per file
==============================
+ analyzer/codechecker_analyzer/arg.py  -3
         

Clones added
============
- web/tests/functional/cli_config/test_store_config.py  1
- web/tests/functional/cli_config/test_server_config.py  7
- web/server/codechecker_server/cmd/server.py  1
- web/client/codechecker_client/cmd/store.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)

**Comment by csordasmarton** [comment]
@gyorgy I already commented on that issue. We can also remove it later if we do not want to keep that.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/tests/functional/cli_config/__init__.py  1
- web/tests/functional/cli_config/test_store_config.py  3
- web/tests/functional/cli_config/test_server_config.py  3
         

Clones added
============
- web/tests/functional/cli_config/test_store_config.py  1
- web/server/codechecker_server/cmd/server.py  1
- web/client/codechecker_client/cmd/store.py  1
- web/tests/functional/cli_config/test_server_config.py  7
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)

**Comment by csordasmarton** [comment]
@gyorb Okay, I have removed this `enabled` key from this configuration file and from the source code.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/tests/functional/cli_config/__init__.py  1
- web/tests/functional/cli_config/test_server_config.py  3
- web/tests/functional/cli_config/test_store_config.py  3
         

Clones added
============
- web/tests/functional/cli_config/test_server_config.py  7
- web/tests/functional/cli_config/test_store_config.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)

File: docs/web/user_guide.md [hunk 9]
```diff
@@ -194,6 +195,17 @@ optional arguments:
...
                         and c/y.cpp. If multiple prefix is given, the longest
                         match will be removed.
+  --config CONFIG_FILE  Allow the configuration from an explicit JSON based
+                        configuration file. The values configured in the
+                        config file will overwrite the values set in the
+                        command line. The format of configuration file is:
+                        {
+                          "store": [
+                            "--name=run_name",
+                            "--tag=my_tag"
```
Comment (gyorb) [inline docs/web/user_guide.md:205]:
```suggestion
                            "--tag=my_tag",
```
A comma is missing, the json is invalid without it.

File: web/client/codechecker_client/cmd/store.py [hunk 10]
```diff
@@ -151,6 +153,22 @@ def add_arguments_to_parser(parser):
...
+                        required=False,
+                        help="R|Allow the configuration from an explicit JSON "
+                             "based configuration file. The values configured "
+                             "in the config file will overwrite the values "
+                             "set in the command line. The format of "
+                             "configuration file is:\n"
+                             "{\n"
+                             "  \"store\": [\n"
+                             "    \"--name=run_name\",\n"
+                             "    \"--tag=my_tag\"\n"
```
Comment (gyorb) [inline web/client/codechecker_client/cmd/store.py:167]:
```suggestion
                             "    \"--tag=my_tag\",\n"
```

File: docs/web/user_guide.md (hunk 9 above)
Comment (csordasmarton) [reply docs/web/user_guide.md]: Good catch, thanks!

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- web/tests/functional/cli_config/__init__.py  1
- web/tests/functional/cli_config/test_store_config.py  3
- web/tests/functional/cli_config/test_server_config.py  3
         

Clones added
============
- web/client/codechecker_client/cmd/store.py  1
- web/tests/functional/cli_config/test_store_config.py  1
- web/tests/functional/cli_config/test_server_config.py  7
- web/server/codechecker_server/cmd/server.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=5173412&bid=16982502)
